include/shared_memory_cache_helper.cuh

```diff
@@ -11,65 +13,225 @@ namespace quda
+    }
+
+    __device__ __host__ inline T load_detail(int x, int y, int z)
+    {
+      atom_t tmp[n_element];
+      int j = (z * block.y + y) * block.x + x;
+#pragma unroll
+      for (int i = 0; i < n_element; i++) tmp[i] = cache<dynamic>()[i * stride + j];
+      T a;
+      memcpy((void*)&a, tmp, sizeof(T));
+      return a;
+    }
+
+    /**
+       @brief Dummy instantiation for the host compiler
+    */
+    template <bool is_device, typename dummy = void> struct sync_impl { void operator()() { } };
+
+    /**
+       @brief Synchronize the cache when on the device
+    */
+    template <typename dummy> struct sync_impl<true, dummy> { __device__ inline void operator()() { __syncthreads(); } };
+
+  public:
+    /**
+       @brief constructor for SharedMemory cache.  If no arguments are
+       pass, then the dimensions are set according to the templates
+       block_size_y and block_size_z, together with the derived
+       block_size_x.  Otherwise use the block sizes passed into the
+       constructor.
+
+       @param[in] block Block dimensions for the 3-d shared memory object 
+    */
+    constexpr SharedMemoryCache(dim3 block = dim3(block_size_x, block_size_y, block_size_z)) :
+      block(block),
+      stride(block.x * block.y * block.z) {}
+
+    /**
+       @brief Grab the raw base address to shared memory.
+    */
+    __device__ __host__ inline T* data() { return reinterpret_cast<T*>(cache<dynamic>()); }
+
+    /**
+       @brief Save the value into the 3-d shared memory cache.
+       Implicitly store at coordinates given by thread_idx().
+       @param[in] a The value to store in the shared memory cache
+     */
+    __device__ __host__ inline void save(const T &a)
     {
-      extern __shared__ int cache_[];
-      return reinterpret_cast<real *>(cache_);
+      save_detail(a, target::thread_idx().x, target::thread_idx().y, target::thread_idx().z);
     }
 
-public:
     /**
-       @brief Save the vector into the 3-d shared memory cache.
-       Implicitly store the vector at coordinates given by threadIdx.
-       @param[in] a The vector to store in the shared memory cache
+       @brief Load a value from the shared memory cache
+       @param[in] x The x index to use
+       @param[in] y The y index to use
+       @param[in] z The z index to use
+       @return The value at coordinates (x,y,z)
      */
-    __device__ inline void save(const Vector &a)
+    __device__ __host__ inline T load(int x = -1, int y = -1, int z = -1)
     {
-      int j = (threadIdx.z * blockDim.y + threadIdx.y) * blockDim.x + threadIdx.x;
-#pragma unroll
-      for (int i = 0; i < 2 * a.size; i++) {
-        cache()[j] = *(reinterpret_cast<const real *>(a.data) + i);
-        j += blockDim.z * blockDim.y * blockDim.x;
-      }
+      auto tid = target::thread_idx();
+      x = (x == -1) ? tid.x : x;
+      y = (y == -1) ? tid.y : y;
+      z = (z == -1) ? tid.z : z;
```

Comment:
Will this produce unwanted checks in kernels?

Reply:
For most cases, the compiler will know these values, so this comparison will be evaluated at compile time, e.g., when we have an empty constructor.

Reply:
But when someone actually specify the values for `x`, `y` or `z` the compiler might perform the comparison. But in that case I guess one can use `load_detail` directly.

Reply:
Yes, and when we have the specialized shifts in a given direction.